File: meetingRoomsII_253.py
# https://leetcode.com/problems/meeting-rooms-ii/

# // Time Complexity : O(N logN)
# // Space Complexity : O(N)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No

# Your approach:
#sort intervals based on starting time
#we only need to take care that the start time of meeting is greater than end time of previous meeting: if it isn't -> add another meeting room otherwise replace the end time of meeting room in the minheap that maintains end times of meetings

import heapq as hq
class Solution:
    def minMeetingRooms(self, intervals: List[List[int]]) -> int:
        
        if not intervals:
            return 0
        
        heap=[]
        intervals.sort()
        
        hq.heappush(heap, intervals[0][1])
        
        for i in intervals[1:]:
            if heap[0]<=i[0]:
                hq.heappop(heap)
            hq.heappush(heap, i[1])
            
        return len(heap)

File: Meeting Rooms II-253.java
// Time Complexity: O(n)
// Space Complexity: O(n)
class Solution {
    public int minMeetingRooms(int[][] intervals) {
        Arrays.sort(intervals, (a,b) -> a[0] - b[0]);
        PriorityQueue<Integer> pq = new PriorityQueue<>();
          for(int i=0;i<intervals.length;i++)
          {  int[] interval = new int[intervals[i].length];
             interval = intervals[i];
            if(pq.size() > 0 && pq.peek() <= interval[0]){
                pq.poll();
            }
            pq.add(interval[1]);
        }
        return pq.size();
    }
        
    }
